distributed/shuffle/shuffle_scheduler.py

```diff
@@ -0,0 +1,217 @@
+        ), f"Output {key} complete; nout_left = {state.out_tasks_left} for shuffle {id}"
+
+        state.out_tasks_left -= 1
+
+        ts: TaskState = self.scheduler.tasks[key]
+        assert (
+            len(ts._worker_restrictions) == 1
+        ), f"Output {key} missing worker restrictions"
+        ts._worker_restrictions.clear()
+        del self.output_keys[key]
+
+        if not state.out_tasks_left:
+            # Shuffle is done. Yay!
+            del self.shuffles[id]
+
+    def erred(self, id: ShuffleId, key: str) -> None:
+        "Handle any task for a shuffle erroring"
+        try:
+            state = self.shuffles.pop(id)
+        except KeyError:
+            return
+
+        if state.barrier_reached:
+            # Remove worker restrictions for output tasks, in case the shuffle is re-submitted
+            for k, id_ in list(self.output_keys.items()):
+                if id_ == id:
+                    ts: TaskState = self.scheduler.tasks[k]
+                    ts._worker_restrictions.clear()
+                    del self.output_keys[k]
+
+    def transition(self, key: str, start: str, finish: str, *args, **kwargs):
+        "Watch transitions for keys we care about"
+        parts = parse_key(key)
+        if parts and len(parts) == 3:
+            prefix, group, id = parts
+
+            if prefix == TASK_PREFIX:
+                if start == "waiting" and finish in ("processing", "memory"):
+                    # transfer/barrier starting to run
+                    if group == "transfer":
+                        return self.transfer(ShuffleId(id), key)
+                    if group == "barrier":
+                        return self.barrier(ShuffleId(id), key)
+
+                # transfer/barrier task erred
+                elif finish == "erred":
+                    return self.erred(ShuffleId(id), key)
+
+        # Task completed
+        if start in ("waiting", "processing") and finish in (
+            "memory",
+            "released",
+            "erred",
+        ):
+            try:
+                id = self.output_keys[key]
+            except KeyError:
+                return
+            # Known unpack task completed or erred
+            if finish == "erred":
+                return self.erred(id, key)
+            return self.unpack(id, key)
+
+    def worker_for_key(self, key: str, npartitions: int, workers: list[str]) -> str:
+        "Worker address this task should be assigned to"
+        # Infer which output partition number this task is fetching by parsing its key.
+        # We have to parse keys, instead of generating the list of expected keys, because
+        # blockwise fusion means they won't just be `shuffle-unpack-abcde`.
+        # FIXME this feels very hacky/brittle.
+        # For example, after `df.set_index(...).to_delayed()`, you could create
+        # keys that don't have indices in them, and get fused (because they should!).
+        m = re.match(r"\(.+, (\d+)\)$", key)
```

Comment:
1. perf nitpick

```python
pattern = re.compile("<pattern>")  # global var

def foo(...):
    pattern.match(key)
```

2. I might want to see a unit test for this. I generally don't trust regular expressions, even though this one looks straight forward...
3. I don't feel comfortable with using such a logic at all. I don't think we should parse keys to infer logic. We do control them but this is not a well defined API and it is very hard to control whether or not keys are mutated at some point in time. The fusing you mentioned is the best example. Any kind of optimization or reordering would have the potential to break this contract easily and I think such systems should be disentangled

Reply:
1. I figured `re` caching would handle it, but sure
2. Agreed a test would be good
3. I don't like this logic at all. As noted though, I can't think of any other options _right now_ to have shuffling still work in the face of output task fusion, since HLG annotations are somewhat broken.

Reply:
It's also important to understand that this key-parsing isn't here because this PR uses a scheduler plugin and https://github.com/dask/distributed/pull/5520 doesn't.

This parsing is here because this PR supports Blockwise fusion of output tasks, and the other doesn't. Whether or not we used a scheduler plugin, the only way I can come up with right now to support fusion of output tasks is to parse key names.

Annotations are the ideal, proper way to deal with this. But if you use annotations, then fusion won't happen, so it defeats the purpose. Therefore, if there's no way to attach additional information to the graph (and there's no way to embed information in the tasks themselves and have it transmitted at runtime to the scheduler, because by the time the task is running, it's already too late), the only information you have is the key names.